### DISCUSSION TITLE
the future of import

### DISCUSSION THREAD

**Discussion by mtoy-googly-moogly** [malloydata/malloy#503] [Developing Malloy]
This is just a placeholder.

Currently the import statement looks like:

`import "string"`

The `"string"` is a URL, which could also be a filename relative to the file which contains the import statement (the most common case), but it really is a URL.

What it does is compile that file, and merge the namespace from that compilation in the current file's namespace.

This is pretty much the way that Javascript works.

We are missing a few of features though, and the intention is to do them eventually:

1) Some way for a file to export less than all of itself
2) Some way to import less than all of a file
3) Some way to manage namespaces in case I need to import from two places which use the same name

**Comment by micahcatlin** [comment]
As a human reader of the written malloy text, I would love to be able to encounter a novel 'token' in the file, and quickly look up to the top of the file and see where it is imported from.   If there is a large block of 'import' statements, any of which could have added to the symbol table, I have to look in all of them.

I understand that may not be idiomatic though, if other symbols appear magically in the table as a result of database schema discovery.    But at least for the symbols that are defined in the malloy language itself, explicitly naming imports and exports seems like a benefit to me.

**Reply by christopherswenson** [reply]
I agree. I would love to see `import { flights } from "./flights.malloy"` (or similar)

**Reply by anikaks** [reply]
The thing I like about that is selective import; having some but not all sources come through seems nice, though I guess my alternative is to split up the upstream file.

**Reply by mtoy-googly-moogly** [reply]
Yeah this is not hard, and it will happen some day. It just needs to pop to the top of the work stack. Splitting things into smaller files is a good workaround until then.

**Reply by mtoy-googly-moogly** [reply]
I am pretty sure @micahcatlin is suggesting that we not import into the default namespace, but rather do something so that when you reference an imported object, the reference itself is clearly imported and not from this file ... something lie this

```
import as faa from "faa.malloy"

source: my_fights is faa.flights + {
  // my refinements
}
```

**Reply by mtoy-googly-moogly** [reply]
JS allows you to have a hierarchical namespace, like above, OR to import into the current namespace, which works much better when the file you are importing only has one thing in it

```
import { makeThing } from "make-thing"
const aThing = makeThing()
```

Instead of 

```
import { * as MakeThing } from "make-thing"
const aThing = MakeThing.makeThing()
```

And this is interesting for malloy because I think "a file which contains only one thing" is pretty common in Malloy. So all this is spinning in my head, not sure what I think the answer is yet.

**Reply by mtoy-googly-moogly** [reply]
Also, just for the record, the IDE will take you to the definition of a thing if it is imported.

**Comment by mtoy-googly-moogly** [comment]
Selective import, and renaming import is in 4.0.
